overlord/devicestate/remodel_test.go

```diff
@@ -0,0 +1,704 @@
+		Model:           "my-model",
+		Serial:          "serialserialserial",
+		SessionMacaroon: "session",
+	})
+	c.Assert(err, IsNil)
+
+	expectedDevice := &auth.DeviceState{
+		Brand:           "my-brand",
+		Model:           "my-model",
+		Serial:          "serialserialserial",
+		SessionMacaroon: "session",
+	}
+
+	var device1 *auth.DeviceState
+	c.Assert(chg.Get("device", &device1), IsNil)
+	c.Check(device1, DeepEquals, expectedDevice)
+
+	device, err = devBE.Device()
+	c.Assert(err, IsNil)
+	c.Check(device, DeepEquals, expectedDevice)
+}
+
+func (s *remodelLogicSuite) TestRemodelDeviceBackendIsolation(c *C) {
+	oldModel := fakeRemodelingModel(nil)
+	newModel := fakeRemodelingModel(map[string]interface{}{
+		"store":    "my-other-store",
+		"revision": "1",
+	})
+
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	// we have a device state
+	devicestatetest.SetDevice(s.state, &auth.DeviceState{
+		Brand:  "my-brand",
+		Model:  "my-model",
+		Serial: "serialserialserial",
+	})
+
+	remodCtx, err := devicestate.RemodelCtx(s.state, oldModel, newModel)
+	c.Assert(err, IsNil)
+
+	chg := s.state.NewChange("remodel", "...")
+
+	err = remodCtx.Init(chg)
+	c.Assert(err, IsNil)
+
+	devBE := devicestate.RemodelDeviceBackend(remodCtx)
+
+	err = devBE.SetDevice(&auth.DeviceState{
+		Brand:           "my-brand",
+		Model:           "my-model",
+		Serial:          "serialserialserial",
+		SessionMacaroon: "session",
+	})
+	c.Assert(err, IsNil)
+
+	// the global device state is as before
+	expectedGlobalDevice := &auth.DeviceState{
+		Brand:  "my-brand",
+		Model:  "my-model",
+		Serial: "serialserialserial",
+	}
+
+	device, err := s.mgr.StoreContextBackend().Device()
+	c.Assert(err, IsNil)
+	c.Check(device, DeepEquals, expectedGlobalDevice)
+}
+func (s *remodelLogicSuite) TestNewStoreRemodelContextStore(c *C) {
+	oldModel := fakeRemodelingModel(nil)
+	newModel := fakeRemodelingModel(map[string]interface{}{
+		"store":    "my-other-store",
+		"revision": "1",
+	})
+
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	// we have a device state
+	devicestatetest.SetDevice(s.state, &auth.DeviceState{
+		Brand:           "my-brand",
+		Model:           "my-model",
+		Serial:          "serialserialserial",
+		SessionMacaroon: "prev-session",
+	})
+
+	remodCtx, err := devicestate.RemodelCtx(s.state, oldModel, newModel)
+	c.Assert(err, IsNil)
+
+	c.Check(s.capturedDevBE, NotNil)
+
+	sto := remodCtx.Store()
+	c.Check(sto, Equals, s.dummyStore)
+
+	s.dummyStore = nil
+
+	sto1 := remodCtx.Store()
+	c.Check(sto1, Equals, sto)
+}
+
+func (s *remodelLogicSuite) TestNewStoreRemodelContextFinish(c *C) {
+	oldModel := fakeRemodelingModel(nil)
+	newModel := fakeRemodelingModel(map[string]interface{}{
+		"store":    "my-other-store",
+		"revision": "1",
+	})
+
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	// we have a device state
+	devicestatetest.SetDevice(s.state, &auth.DeviceState{
+		Brand:  "my-brand",
+		Model:  "my-model",
+		Serial: "serialserialserial",
```

Comment:
I think I would find it easier if `		SessionMacaroon: "prev-session",` was set here and below in the new device we would use something like "SessionMacaroon: "new-store-session"" (or something). Assuming I read this correctly it would have helped me to grasp the details easier.